specs/transport/car/carv2/index.md

```diff
@@ -0,0 +1,177 @@
+
+1. **Characteristics**: A 128-bit (16-byte) bitfield used to describe certain features of the enclosed data.
+2. **Data offset**: A 64-bit (8-byte) unsigned little-endian integer indicating the byte-offset from the beginning of the container to the first byte of the CARv1 data payload.
+3. **Data length**: A 64-bit (8-byte) unsigned little-endian integer indicating the byte-length of the CARv1 data payload.
+4. **Index offset**: A 64-bit (8-byte) unsigned little-endian integer indicating the byte-offset from the beginning of the container to the first byte of the index payload.
+
+### Characteristics
+
+The characteristics bitfield contained within the CARv2 header may be used to indicate certain features of the specific CARv2 container.
+
+**TODO:** spit-balling here, these feel like feature-bloat so we should probably only add features as we use them but should be clear to leave enough space to encode additional things we care about without causing conflicts - i.e. it shouldn't be possible to flip bits to say that it's _both_ depth-first and breadth-first. Hence the suggestion to use easy integers where there's >2 options.
+
+1. **Bits 0 to 7**: read as an unsigned little-endian 8-bit integer describing the ordering of the blocks in the CARv1 payload:
+   * `0` indicates _no particular order_, no assumptions should be made regarding determinism of payload layout.
+   * `1` indicates a _depth-first_ ordering of blocks in the DAG(s) referenced by the "roots" array. (**TODO:** "depth-first" isn't enough to be clear about what this means for a DAG walk - e.g. do we yield the current block before going to the next, or wait till we reach the depth before yielding blocks?)
+   * `2` indicates a _breadth-first_ ordering of blocks in the DAG(s) referenced by the "roots" array. (**TODO:** ditto ^ what does this specifically mean in DAG terms)
+   * `3` indicates that the ordering is dictated by an [IPLD Selector](../specs/selectors/index.md) which itself is serialized into the CARv1 payload and referenced in the last element in the "roots" array. i.e. deserializing the block referenced by the last element of the roots array in the CARv1 header should yield a data model form of the IPLD selector that was used to select and order the blocks in the CARv1 payload.
+   * `4` has the same meaning as `3` except that the IPLD selector is encoded as a DAG-CBOR IPLD block immediately following the Header and is prefixed with a varint indicating the length of the bytes to decode.
+2. **Bit 8** indicating whether the blocks have been properly deduplicated (`1`) and that a reader should not expect any repeated CIDs, or that the blocks _may_ contain duplicates (`0`) (perhaps as a trade-off for speed or memory consumption during generation).
+
+Future amendments of this specification may introduce additional characteristics, including:
+
+* **TODO:** what else are we potentially filling 128-bits up with?
```

Comment:
I thought we were leaving the characteristics bits unused for now :) Your two ideas sound interesting here, though I wonder if we need them right now in the spec, given that the first version of go-car v2 won't support them.

Reply:
I mainly ask this because 128-bits is a lot of bits and I'd like to see what's behind wanting so much optionality here. This TODO is under "Future ammendments ... including:", if we can't come up with many ideas then maybe we need to do some careful thinking about this field.